### ISSUE TITLE
Improve 8-bit quantization support

### ISSUE THREAD

**Issue by lantiga** [Lightning-AI/lit-llama#121]
As per https://github.com/TimDettmers/bitsandbytes#tldr

- in generate.py pass `has_fp16_weights=False` (the fact we don't do this might explain the fact that memory requirements for 8-bit quantization are currently similar to `bfloat16`) to `Linear8bitLt`
- for finetuning, pass `has_f16_weights=True` (the default) and `threshold=6.0` 
- Introduce 8-bit optimizers for fine-tuning

Note that we need to derive intermediate classes from `Linear8bitLt` that have `has_fp16_weights` and `threshold` set to the appropriate values, so we can set them as Linear substitutes in the context manager.

**Comment by KnutJaegersberg** [comment]
Is 4-bit fine tuning possible, too? We could then fine tune a 30B model on an RTX 3090.

**Comment by lantiga** [comment]
@KnutJaegersberg we actually already support fine-tuning on a single 3090 at 16 bit precision, which is quite nice: https://github.com/Lightning-AI/lit-llama#finetune-the-model

but we'll keep looking at ways to shrink further, we'll start from 8 bit and keep 4 bit for inference. But keeping the frozen weights in 4 bit and the LoRA / Adapter weights 16 bit is an idea we want to pursue.

**Comment by awaelchli** [comment]
@lantiga For the last point:

> Introduce 8-bit optimizers for fine-tuning

I tried that early on and it didn't make a big difference, since in lora and adapters the number of trainable weights is so small, the savings in optimizer state is very minimal. But I can get the exact numbers.

**Comment by lantiga** [comment]
One question (discussed offline, reporting here for visibility) is also: can we fine tune lora and adapters 
- keeping everything 8 bit (or 4bit - although now we have a performance hit from doing that, this is on @t-vi's radar)
- keeping the frozen model 8bit (or 4bit) and the adapter / lora layers bfloat16

**Comment by awaelchli** [comment]
For the record, had already made a quick experiment doing the latter point (branch here https://github.com/Lightning-AI/lit-llama/compare/adapter-quantized?expand=1) but the loss wouldn't converge and it impacted the speed, I didn't yet investigate further.

**Comment by rcmalli** [comment]
I have seen that after linear layers are initialized using `bitsandbytes` there are some fixes for stability improvements. Here is a small snippet taken from [this notebook](https://colab.research.google.com/drive/1jCkpikz0J2o20FBQmYmAGdiKmJGOMo-o?usp=sharing#scrollTo=T-gy-LxM0yAi). It might be helpful.

```python
for param in model.parameters():
  param.requires_grad = False  # freeze the model - train adapters later
  if param.ndim == 1:
    # cast the small parameters (e.g. layernorm) to fp32 for stability
    param.data = param.data.to(torch.float32)
class CastOutputToFloat(nn.Sequential):
  def forward(self, x): return super().forward(x).to(torch.float32)
model.lm_head = CastOutputToFloat(model.lm_head)
```

**Comment by t-vi** [comment]
@rcmalli Interesting thank you!
I think if we do something like that, we should be casting things back somewhere, too?
Also, if we use bf16, is the casting needed as urgently?
